// is-what-type/dist/index.mjs
import { default as default2 } from "./getType.mjs";
import { default as default3 } from "./isAbsoluteURL.mjs";
import { default as default4 } from "./isArrayLike.mjs";
import { default as default5 } from "./isAsyncFunction.mjs";
import { default as default6 } from "./isESModule.mjs";
import { default as default7 } from "./isError.mjs";
import { default as default8 } from "./isFunction.mjs";
import { default as default9 } from "./isIterable.mjs";
import { default as default10 } from "./isLeapYear.mjs";
import { default as default11 } from "./isLength.mjs";
import { default as default12 } from "./isNil.mjs";
import { default as default13 } from "./isNumber.mjs";
import { default as default14 } from "./isObject.mjs";
import { default as default15 } from "./isPlainObject.mjs";
import { default as default16 } from "./isPromiseLike.mjs";
import { default as default17 } from "./isValidDate.mjs";
import { default as default18 } from "./isWhat.mjs";
export {
  default2 as getType,
  default3 as isAbsoluteURL,
  default4 as isArrayLike,
  default5 as isAsyncFunction,
  default6 as isESModule,
  default7 as isError,
  default8 as isFunction,
  default9 as isIterable,
  default10 as isLeapYear,
  default11 as isLength,
  default12 as isNil,
  default13 as isNumber,
  default14 as isObject,
  default15 as isPlainObject,
  default16 as isPromiseLike,
  default17 as isValidDate,
  default18 as isWhat
};
